inference-engine/src/mkldnn_plugin/nodes/bucketize.cpp

```diff
@@ -72,46 +77,99 @@ class BucketizeImpl : public ExtLayerBase {
+            bucketize<int, int64_t, int>(inputs[0], inputs[1], outputs[0]);
+        } else if (input_precision == Precision::I32 && boundaries_precision == Precision::I64 &&
+            output_precision == Precision::I64) {
+            bucketize<int, int64_t, int64_t>(inputs[0], inputs[1], outputs[0]);
+        } else if (input_precision == Precision::I64 && boundaries_precision == Precision::FP32 &&
+            output_precision == Precision::I32) {
+            bucketize<int64_t, float, int>(inputs[0], inputs[1], outputs[0]);
+        } else if (input_precision == Precision::I64 && boundaries_precision == Precision::FP32 &&
+            output_precision == Precision::I64) {
+            bucketize<int64_t, float, int64_t>(inputs[0], inputs[1], outputs[0]);
+        } else if (input_precision == Precision::I64 && boundaries_precision == Precision::I32 &&
+            output_precision == Precision::I32) {
+            bucketize<int64_t, int, int>(inputs[0], inputs[1], outputs[0]);
+        } else if (input_precision == Precision::I64 && boundaries_precision == Precision::I32 &&
+            output_precision == Precision::I64) {
+            bucketize<int64_t, int, int64_t>(inputs[0], inputs[1], outputs[0]);
+        } else if (input_precision == Precision::I64 && boundaries_precision == Precision::I64 &&
+            output_precision == Precision::I32) {
+            bucketize<int64_t, int64_t, int>(inputs[0], inputs[1], outputs[0]);
+        } else if (input_precision == Precision::I64 && boundaries_precision == Precision::I64 &&
+            output_precision == Precision::I64) {
+            bucketize<int64_t, int64_t, int64_t>(inputs[0], inputs[1], outputs[0]);
+        } else {
+            return GENERAL_ERROR;
         }
-        int *output_tensor_ptr = outputs[OUTPUT_TENSOR_PORT]->cbuffer().as<int *>() +
-            inputs[OUTPUT_TENSOR_PORT]->getTensorDesc().getBlockingDesc().getOffsetPadding();
+
+        return OK;
+    }
+
+private:
+    template <typename T, typename T_BOUNDARIES, typename T_IND>
+    void bucketize(Blob::Ptr input, Blob::Ptr boundaries, Blob::Ptr output) {
+        const auto *input_data = input->cbuffer().as<const T *>();
+        const auto *boundaries_data = input->cbuffer().as<const T_BOUNDARIES *>();
+        auto *output_data = output->buffer().as<T_IND *>();
 
         if (with_bins == false) {
             for (size_t ind = 0; ind < num_values; ind++) {
-                output_tensor_ptr[ind] = 0;
+                output_data[ind] = 0;
             }
-            return OK;
+            return;
         }
 
         for (size_t ind = 0; ind < num_values; ind++) {
-            float value = input_tensor_ptr[ind];
+            T value = input_data[ind];
 
             // find a bin to which value belongs
-            output_tensor_ptr[ind] = -1;
+            output_data[ind] = static_cast<T_IND>(num_bin_values);
             for (size_t bin_ind = 0; bin_ind < num_bin_values; bin_ind++) {
```

Comment:
Since the bins boundaries are ordered you can use binary search to find the index of the bin. This will speedup the algorithm from O(N * N) to O(N * log(N)). Take a look at `lower_bound` or `upper_bound` STL methods.

Reply:
Done